Composite/Composite/Transforms.fs

```diff
@@ -57,12 +57,25 @@ module Transforms =
         getPages inSeq
 
     let toBatched batchSize getElemSize inSeq =
-        let rec getBatchAndRest numRemains pageAndRest = 
-            match numRemains > 0, pageAndRest with
-            | true, (p, r) -> match r |> Seq.tryHead with
-                                | Some h -> getBatchAndRest (numRemains-(getElemSize h)) (Array.append p [|h|], r |> Seq.tail)
-                                | None -> pageAndRest
-            | false, pr -> pr
+        // Here we extract a batch from an input sequence.
+        // We do this recursively, passing the amount of remaining
+        // space in batch as a parameter, plus a tuple comprising
+        // the batch under constuction and the remaining sequence
+        // we have to walk through.
+        let rec getBatchAndRest roomLeft batchAndRest = 
+            let (page, rest) = batchAndRest
+            rest
+            |> Seq.tryHead
+            |> function
+               | Some head -> let size = getElemSize head
+                              let roomLeftNew = roomLeft - size
+                              page
+                              |> function
+                                 | [||] -> getBatchAndRest roomLeftNew ([|head|], rest |> Seq.tail)

```

Comment:
In the case when a weight of the first element in a batch greater than a batch size it will be added to a batch without any checks.

Reply:
@v-ilin This is precisely my idea here. A batch must contain at least one element, doesn't matter how heavy. Otherwise, it would only circle around in a meaningless loop. A multiple element batch cannot be of size greater than the provided batch size.